node/core/dispute-participation/src/tests.rs

```diff
@@ -0,0 +1,466 @@
+		number: block_number,
+		digest: Default::default(),
+		state_root: Default::default(),
+		extrinsics_root: Default::default(),
+	};
+
+	let block_hash = block_header.hash();
+
+	virtual_overseer
+		.send(FromOverseer::Signal(OverseerSignal::ActiveLeaves(
+			ActiveLeavesUpdate::start_work(ActivatedLeaf {
+				hash: block_hash,
+				span: Arc::new(jaeger::Span::Disabled),
+				number: block_number,
+				status: LeafStatus::Fresh,
+			}),
+		)))
+		.await;
+}
+
+async fn participate(virtual_overseer: &mut VirtualOverseer) {
+	let commitments = CandidateCommitments::default();
+	let candidate_receipt = {
+		let mut receipt = CandidateReceipt::default();
+		receipt.commitments_hash = commitments.hash();
+		receipt
+	};
+	let candidate_hash = candidate_receipt.hash();
+	let session = 1;
+	let n_validators = 10;
+
+	virtual_overseer
+		.send(FromOverseer::Communication {
+			msg: DisputeParticipationMessage::Participate {
+				candidate_hash,
+				candidate_receipt: candidate_receipt.clone(),
+				session,
+				n_validators,
+			},
+		})
+		.await;
+}
+
+async fn recover_available_data(virtual_overseer: &mut VirtualOverseer) {
+	let pov_block = PoV {
+		block_data: BlockData(Vec::new()),
+	};
+
+	let available_data = AvailableData {
+		pov: Arc::new(pov_block),
+		validation_data: Default::default(),
+	};
+
+	assert_matches!(
+		virtual_overseer.recv().await,
+		AllMessages::AvailabilityRecovery(
+			AvailabilityRecoveryMessage::RecoverAvailableData(_, _, _, tx)
+		) => {
+			tx.send(Ok(available_data)).unwrap();
+		},
+		"overseer did not receive recover available data message",
+	);
+}
+
+async fn fetch_validation_code(virtual_overseer: &mut VirtualOverseer) {
+	let validation_code = ValidationCode(Vec::new());
+
+	assert_matches!(
+		virtual_overseer.recv().await,
+		AllMessages::RuntimeApi(RuntimeApiMessage::Request(
+			_,
+			RuntimeApiRequest::ValidationCodeByHash(
+				_,
+				tx,
+			)
+		)) => {
+			tx.send(Ok(Some(validation_code))).unwrap();
+		},
+		"overseer did not receive runtime api request for validation code",
+	);
+}
+
+async fn store_available_data(virtual_overseer: &mut VirtualOverseer, success: bool) {
+	assert_matches!(
+		virtual_overseer.recv().await,
+		AllMessages::AvailabilityStore(AvailabilityStoreMessage::StoreAvailableData(
+			_,
+			_,
+			_,
+			_,
+			tx,
+		)) => {
+			if success {
+				tx.send(Ok(())).unwrap();
+			} else {
+				tx.send(Err(())).unwrap();
+			}
+		},
+		"overseer did not receive store available data request",
+	);
+}
+
+#[test]
+fn cannot_participate_when_recent_block_state_is_missing() {
+	test_harness(|mut virtual_overseer| {
+		Box::pin(async move {
+			participate(&mut virtual_overseer).await;
+
+			virtual_overseer
+				.send(FromOverseer::Signal(OverseerSignal::Conclude))
+				.await;
+
+			// no request is received by the overseer which means that no
+			// attempt to participate was made
+			assert!(virtual_overseer.try_recv().await.is_none());
```

Comment:
we could put this logic into test_harness like we do in https://github.com/paritytech/polkadot/blob/1dfc23f61eea51bfaad69ce0723ac56bf4b11572/node/network/gossip-support/src/tests.rs#L54-L61

to avoid duplication

Reply:
Updated 👍